drgn_tools/virtutil.py

```diff
@@ -4,13 +4,206 @@
+    for vm in vm_list:
+        if has_member(vm, "vcpus"):
+            struct_vcpu = vm.vcpus
+            for i in range(len(struct_vcpu)):
+                if hex(struct_vcpu[i].value_()) == "0x0":
+                    break
+                vcpu = hex(struct_vcpu[i].value_())
+                id = struct_vcpu[i].vcpu_id.value_()
+                if has_member(struct_vcpu[i], "vcpu_idx"):
+                    idx = struct_vcpu[i].vcpu_idx.value_()
+                else:
+                    idx = i
+                arch = hex(struct_vcpu[i].arch.address_of_())
+                stat = hex(struct_vcpu[i].stat.address_of_())
+                if has_member(struct_vcpu[i], "stats_id"):
+                    stat_id = struct_vcpu[i].stats_id.string_().decode("utf-8")
+                else:
+                    # The UEK5 does not have kvm_vcpu.stats_id structure member. Print None instead.
+                    stat_id = None
+                state = struct_vcpu[i].arch.mp_state.value_()
+                cpu = struct_vcpu[i].cpu.value_()
+                if has_member(struct_vcpu[i], "wait"):
+                    task = hex(struct_vcpu[i].wait.task.value_())
+                else:
+                    # The UEK5 does not have kvm_vcpu.wait structure member. Print None instead.
+                    task = None
+                rows.append(
+                    [
+                        hex(vm.value_()),
+                        vcpu,
+                        id,
+                        idx,
+                        arch,
+                        stat,
+                        stat_id,
+                        KVM_VCPU_STATE[state],
+                        cpu,
+                        str(task),
+                    ]
+                )
+        # For UEK-NEXT:
+        else:
+            for _, entry in xa_for_each(vm.vcpu_array.address_of_()):
+                struct_vcpu = cast("struct kvm_vcpu *", entry)
+                if hex(struct_vcpu.value_()) == "0x0":
+                    break
+                vcpu = hex(struct_vcpu.value_())
+                id = struct_vcpu.vcpu_id.value_()
+                idx = struct_vcpu.vcpu_idx.value_()
+                arch = hex(struct_vcpu.arch.address_of_())
+                stat = hex(struct_vcpu.stat.address_of_())
+                stat_id = struct_vcpu.stats_id.string_().decode("utf-8")
+                state = struct_vcpu.arch.mp_state.value_()
+                cpu = struct_vcpu.cpu.value_()
+                task = hex(struct_vcpu.wait.task.value_())
+
+                rows.append(
+                    [
+                        hex(vm.value_()),
+                        vcpu,
+                        id,
+                        idx,
+                        arch,
+                        stat,
+                        stat_id,
+                        KVM_VCPU_STATE[state],
+                        cpu,
+                        str(task),
+                    ]
+                )
```

Comment:
Rather than having two distinct loops with a lot of duplicate code, we could clean this up. Let's define a `for_each_vcpu()` function like below:

```python
def for_each_vcpu(vm: Object) -> Iterator[Object]:
    if has_member(vm, "vcpus"):
        vcpu_iterator = iter(vm.vcpus)
    else:
        vcpu_iterator = (
            cast("struct kvm_vcpu *", e)
            for _, e in xa_for_each(vm.vcpu_array.address_of_())
        )
    for struct_vcpu in vcpu_iterator:
        if struct_vcpu.value_() == 0:
            break
        yield struct_vcpu
```

Then you can use that helper here to consolidate, so that you don't repeat the printing code. Plus we get a nice helper for others to use.